core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp

```diff
@@ -523,44 +523,55 @@ class ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>,
     int max_active_league_size =
         OpenMPTargetExec::MAX_ACTIVE_THREADS / team_size;
 
-    // A lock array to keep track of scratch memory blocks.
-    // The first element to keep track of the number of free-blocks.
-    int* lock_array = new int[max_active_league_size + 1];
-    memset(lock_array, 1, (max_active_league_size + 1) * sizeof(int));
-    lock_array[0] = max_active_league_size;
+    // Allocate the lock_array on the device and initialize it to 0's in a
+    // target region.
+    int* lock_array = static_cast<int*>(omp_target_alloc(
```

Comment:
This ptr still needs to come from the OpenMPTarget instance, in particular so you don't reallocate it all the time. Only reallocate if it has to grow.